linkerd/core/src/main/scala/io/buoyant/linkerd/RequestActiveTracer.scala

```diff
@@ -0,0 +1,265 @@
+    }
+
+    val addresses = lbSet.map {
+      case Addr.Bound(a, _) => Some(
+        a.map {
+          case inetAddr: Address.Inet => inetAddr.addr.toString.stripPrefix("/")
+          case _ => ""
+        }
+      )
+      case _ => None
+    }
+
+    val dtreeF = namers.interpreter match {
+      case delegator: Delegator => delegator.delegate(dtab.dtab(), serviceName)
+        .map(Some(_))
+      case _ => Future.None
+    }
+
+    dtreeF.joinWith(addresses) {
+      case (dTree: Option[DelegateTree[Name.Bound]], addrSet: Option[Set[String]]) =>
+        val tree = formatDelegation(dTree.getOrElse(EmptyDelegateTree), List.empty, clientPath)
+
+        resp.contentString = formatRouterContext(
+          resp.contentString,
+          serviceName.show,
+          clientPath.show,
+          selectedEndpoint,
+          addrSet,
+          tree.map(_.toString),
+          stopwatch().inMillis.toString
+        )
+
+        //We set the content length of the response in order to view the content string entirely
+        resp.contentLength = resp.content.length
+        resp
+    }
+  }
+
+  override def apply(
+    req: Request,
+    svc: Service[Request, Response]
+  ): Future[Response] = {
+
+    val maxTraceDepth = Try(req.headerMap.get(MaxForwardsHeader).map(_.toInt))
+    val addRouterCtxt = req.headerMap.get(AddRouterContextHeader)
+
+    (req.method, maxTraceDepth, addRouterCtxt) match {
+      case (Method.Trace, Throw(_: NumberFormatException), _) =>
+        // Max-Forwards header is unparseable
+        val resp = Response(Status.BadRequest)
+        resp.contentString = s"Invalid value for $MaxForwardsHeader header"
+        Future.value(resp)
+      case (Method.Trace, Return(Some(0)), isAddRouterCtxt) =>
+        // Max-Forwards header is 0 and the l5d-add-context header is present
+        isAddRouterCtxt match {
+          // returns a response with no router context
+          case None => Future.value(Response(req))
+
+          // returns a response with router context
+          case _ =>
+            val stopwatch = Stopwatch.start()
+            getRequestTraceResponse(Response(req), stopwatch)
+        }
+
+      case (Method.Trace, Return(Some(num)), isAddRouterCtxt) if num > 0 =>
+        // Decrement Max-Fowards header
+        req.headerMap.set(MaxForwardsHeader, (num - 1).toString)
+
+        isAddRouterCtxt match {
+          // Forward requests without adding router context to response
+          case None => svc(req).ensure {
+            req.headerMap.set(MaxForwardsHeader, num.toString); ()
+          }
+          case _ =>
+            // Forward requests and router context to response
+            val stopwatch = Stopwatch.start()
+            svc(req).flatMap((prevResp: Response) => getRequestTraceResponse(prevResp, stopwatch)).ensure {
+              req.headerMap.set(MaxForwardsHeader, num.toString); ()
+            }
+        }
+      case (_, _, _) => svc(req)
```

Comment:
I think if we get a TRACE with l5d-add-context but no Max-Forwards, we should still add the context.  I think we basically treat this as Max-Forwards = infinity.

Reply:
Yea I thought about that as well. Could we run into a scenario were we have Linkerd set up in a linker to linker fashion and when a request with no `Max-Forwards` could eventually make the request hang in a linker to linker loop? I guess the timeout exception might make this a non-issue.

Reply:
Yes, although this is already an issue with Linkerd: https://github.com/linkerd/linkerd/issues/1411